### PR TITLE
chore: refactor searcher operations out of master side searchers

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

refactor of master-side searchers:
- remove usage of searcher operations
- remove usage of `max_length` master-side
- rewrite of preview-search
- renames trial -> run in new code
- rewrite of progress calculation

TODO:
- code cleanup / add comments
- rewrite/add tests
- update docs

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/tests/cli/test_cli.py [hunk 1]
```diff
@@ -566,3 +569,46 @@ def test_dev_bindings_call_arg_unmarshal(case: Tuple[List[str], Dict[str, Any]])
...
+                    params={
+                        "config": searcher_config,
+                    }
+                )
+            ],
+            json=mock_resp.to_json(),
+        )
+        cli.main(
+            ["preview-search", str(conf_path)]
+        )
```
Comment (rb-determined-ai) [inline harness/tests/cli/test_cli.py:614]: missing newline at end of file, please fix your editor to not do this

File: hpc-ard-launcher-go/launcher/go.sum
Comment (rb-determined-ai) [inline hpc-ard-launcher-go/launcher/go.sum:1]: what's going on with this file?  This doesn't seem right...

Comment (azhou-determined) [reply hpc-ard-launcher-go/launcher/go.sum:1]: shoot, yeah, something's wrong with my setup on new computer. removed.

File: harness/tests/cli/util.py [hunk 2]
```diff
@@ -151,3 +153,14 @@ def expect_get_info(
...
+
+def check_cli_output(args: List[str], expected: str) -> None:
+    """
+    Helper method to test CLI methods that checks redirected STDOUT from the executed command
+    matches expected output.
+    """
+    with contextlib.redirect_stdout(io.StringIO()) as f:
+        cli.main(args=args)
+    actual = f.getvalue()
+    assert expected == actual, "CLI output does not match expected output."
```
Comment (rb-determined-ai) [inline harness/tests/cli/util.py:166]: If you are going to make a helper function out of this, you need better error logs.

Try this:
```python
import difflib

if expected == actual:
  return
exp_lines = expected.splitlines(keepends=True)
act_lines = actual.splitlines(keepends=True)
diff_lines = difflib.ndiff(act_lines, exp_lines)
diff = ''.join(diff_lines)
raise ValueError(f"cli output for {args} output actual (-) != expected (+):\n{diff}")

File: master/pkg/searcher/adaptive_asha_test.go [hunk 3]
```diff
@@ -25,194 +27,65 @@ func TestBracketMaxConcurrentTrials(t *testing.T) {
...
-			config: expconf.SearcherConfig{
-				RawSmallerIsBetter: ptrs.Ptr(true),
-				RawAdaptiveASHAConfig: &expconf.AdaptiveASHAConfig{
-					RawMaxLength: ptrs.Ptr(expconf.NewLengthInBatches(900)),
-					RawMaxTrials: ptrs.Ptr(5),
-					RawMode:      modePtr(expconf.StandardMode),
-					RawMaxRungs:  ptrs.Ptr(2),
-					RawDivisor:   ptrs.Ptr[float64](3),
+			conf: expconf.AdaptiveASHAConfig{
+				RawMode:                expconf.AdaptiveModePtr(expconf.StandardMode),
```
Comment (rb-determined-ai) [inline master/pkg/searcher/adaptive_asha_test.go:37]: Oh, this function is really old.  I wrote it before I wrote the generics-based `ptrs.Ptr()` (because we didn't have go generics at the time).

Please find all places where this function is used, and replace them with `ptrs.Ptr()`, please.

File: master/pkg/searcher/search_method.go [hunk 4]
```diff
@@ -15,30 +15,30 @@ type context struct {
...
+	// progress returns search progress as a float between 0.0 and 1.0.
+	progress(map[int32]float64, map[int32]bool) float64
+	// runExitedEarly informs the searcher that the run has exited earlier than expected.
+	runExitedEarly(
+		ctx context, runID int32, exitedReason model.ExitedReason,
+	) ([]Action, error)
 
 	// TODO: refactor as model.Snapshotter interface or something
 	model.Snapshotter
 	expconf.InUnits
```
Comment (rb-determined-ai) [inline master/pkg/searcher/search_method.go:40]: shouldn't we be able to get rid of exconf.InUnits as part of this refactor?

File: master/pkg/searcher/search_method.go [hunk 5]
```diff
@@ -69,38 +69,31 @@ func NewSearchMethod(c expconf.SearcherConfig) SearchMethod {
 		return newRandomSearch(*c.RawRandomConfig)
 	case c.RawGridConfig != nil:
 		return newGridSearch(*c.RawGridConfig)
-	case c.RawAsyncHalvingConfig != nil:
-		if c.RawAsyncHalvingConfig.StopOnce() {
-			return newAsyncHalvingStoppingSearch(*c.RawAsyncHalvingConfig, c.SmallerIsBetter())
-		}
-		return newAsyncHalvingSearch(*c.RawAsyncHalvingConfig, c.SmallerIsBetter())
```
Comment (rb-determined-ai) [inline master/pkg/searcher/search_method.go:76]: Why did you delete the AsyncHalvingConfig case?

File: schemas/expconf/v0/searcher-adaptive-asha.json [hunk 6]
```diff
@@ -33,6 +33,20 @@
             "default": null,
             "minimum": 1
         },
+        "time_metric": {
```
Comment (rb-determined-ai) [inline schemas/expconf/v0/searcher-adaptive-asha.json:36]: should time_metric and max_time be eventuallyRequired?

IIUC we won't accept asha jobs without them.

File: schemas/expconf/v0/searcher-adaptive-asha.json [hunk 7]
```diff
@@ -33,6 +33,20 @@
...
+            "type": [
+                "string",
+                "null"
+            ],
+            "default": null
+        },
+        "max_time": {
```
Comment (rb-determined-ai) [inline schemas/expconf/v0/searcher-adaptive-asha.json:43]: also, shouldn't async_halving also be updated to include time_metric and max_time?

File: schemas/expconf/v0/searcher-adaptive-asha.json (hunk 6 above)
Comment (rb-determined-ai) [reply schemas/expconf/v0/searcher-adaptive-asha.json:36]: hm... actually I'm not quite sure how to handle this one.

I suppose, since we still haven't eradicated all of the legacy expconf parsing (which is why we still allow the EOL searchers in the json-schema in the first place), the safest place for a time_metric and max_time check is actually in:

    pkg/scheams/expconf/searcher_config.go::SearcherConfig.AssertCurrent().

That will prevent dumb bugs like "`det e set-checkpoint-gc-policy` fails for old asha experiments" from happening, or either of the other two or three weird places where we wrongly try to parse an experiment config.

File: master/pkg/searcher/search_method.go (hunk 4 above)
Comment (rb-determined-ai) [reply master/pkg/searcher/search_method.go:40]: actually, I mean that as in "we should remove it from this interface" but I suppose it's actually just pointless now, and can be deleted from the entire codebase, right?

File: master/pkg/searcher/grid.go [hunk 8]
```diff
@@ -36,85 +36,77 @@ func newGridSearch(config expconf.GridConfig) SearchMethod {
...
+		create := NewCreate(ctx.rand, params)
+		actions = append(actions, create)
+		s.PendingRuns++
 	}
-	return ops, nil
+	return actions, nil
 }
 
-func newHyperparameterGrid(params expconf.Hyperparameters) []HParamSample {
+func NewHyperparameterGrid(params expconf.Hyperparameters) []HParamSample {
```
Comment (rb-determined-ai) [inline master/pkg/searcher/grid.go:109]: why suddenly export this?

File: master/pkg/searcher/operations.go [hunk 9]
```diff
@@ -104,76 +102,31 @@ type Requested interface {
 	GetRequestID() model.RequestID
 }
 
-// Create a new trial for the search method.
-type Create struct {
+// LegacyCreate a new trial for the search method.
+type LegacyCreate struct {
```
Comment (rb-determined-ai) [inline master/pkg/searcher/operations.go:106]: delete this?  It seems unused...

Comment (azhou-determined) [reply master/pkg/searcher/operations.go]: yeah, i kept this mainly for reference as i was rewriting logic. deleted.

File: master/pkg/searcher/grid.go (hunk 8 above)
Comment (azhou-determined) [reply master/pkg/searcher/grid.go]: was probably doing some funny things at one point. fixed.

File: schemas/expconf/v0/searcher-adaptive-asha.json (hunk 7 above)
Comment (azhou-determined) [reply schemas/expconf/v0/searcher-adaptive-asha.json:43]: see other comment: i thought `async_halving` was a deprecated searcher.

File: master/pkg/searcher/search_method.go (hunk 5 above)
Comment (azhou-determined) [reply master/pkg/searcher/search_method.go:76]: i thought `async_halving` was a deprecated searcher and has been for a long time, so i removed it. but now i'm not so sure. let me circle back on this.

File: master/pkg/searcher/search_method.go (hunk 4 above)
Comment (azhou-determined) [reply master/pkg/searcher/search_method.go]: yes and yes and done

File: harness/tests/cli/util.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/tests/cli/util.py]: Yeah, because of course python stdlib has a diffing tool.

Don't tell anybody, but there's a complete implementation of determined in the python stdlib too.  It's from the 90's though so nobody remembers to look there.

File: master/pkg/searcher/search_method.go (hunk 5 above)
Comment (rb-determined-ai) [reply master/pkg/searcher/search_method.go:76]: I thought we talked about this?  I thought async_halving is just like adaptive except for it has a manually-chosen time constant.  Or rather, adaptive is like multiple trees of async_halving with different time constants, if you don't happen to know which time constant is right for your model.

File: master/internal/api_trials.go [hunk 10]
```diff
@@ -1525,6 +1456,26 @@ func (a *apiServer) ReportTrialMetrics(
 		experiment.AuthZProvider.Get().CanEditExperiment); err != nil {
 		return nil, err
 	}
+	defer func() {
+		// Notify searcher of validation metrics.
+		if metricGroup == model.ValidationMetricGroup {
```
Comment (rb-determined-ai) [inline master/internal/api_trials.go:1461]: code style suggestion:

EITHER reduce indentation:

```go
if metricGroup != model.ValidationMetricGroup {
  return
}
```

OR don't defer pointless functions:

```go
if metricGroup == model.ValidationMetricGroup {
    defer func() {
    }()
}
```

take your pick.  I like the first because less indentation is good for readability, and I like the second because not calling a noop is also good for readability.

File: master/internal/db/postgres_trial.go [hunk 11]
```diff
@@ -276,17 +276,17 @@ func TrialTaskIDsByTrialID(ctx context.Context, trialID int) ([]*model.RunTaskID
 	return ids, nil
 }
 
-// TrialByExperimentAndRequestID looks up a trial, returning an error if none exists.
-func TrialByExperimentAndRequestID(
-	ctx context.Context, experimentID int, requestID model.RequestID,
-) (*model.Trial, error) {
```
Comment (rb-determined-ai) [inline master/internal/db/postgres_trial.go:282]: Hm... are we guaranteed that requestIDs are not reused between experiments?

A long time ago, we generated UUIDs from the experiment seed, so two experiments with the same seed would actually get identical UUIDs.  Which sounds stupid, but was probably necessary in the earliest version of our searchers, which were built on a sort of writeahead log strategy with replays for restoring.

Based on the fact that model.NewRequestID still takes an RNG I'd say you probably still need the experiment id here, unfortunately.

Comment (rb-determined-ai) [reply master/internal/db/postgres_trial.go:282]: or... actually... is deleting all remaining NewRequestID's on your todo list?  It looks like it isn't used outside of tests, so maybe my feedback here is obsolete.

Comment (rb-determined-ai) [reply master/internal/db/postgres_trial.go:282]: although... you'd probably need to kill all non-terminal experiments to make this change, because right after upgrading, you are still going to have to restore experiments from where the request id wasn't unique.

File: master/pkg/searcher/search_method.go (hunk 5 above)
Comment (azhou-determined) [reply master/pkg/searcher/search_method.go]: yeah, i remember now, you're right. i added it back. and since we're supporting it, i'll make a note to add it to our docs, since we don't officially advertise that we support it.

File: master/internal/api_trials.go (hunk 10 above)
Comment (azhou-determined) [reply master/internal/api_trials.go]: chose the second, read better to me

File: master/internal/db/postgres_trial.go (hunk 11 above)
Comment (azhou-determined) [reply master/internal/db/postgres_trial.go]: gah. yeah, i was afraid this was the case. added back experiment ID

File: harness/determined/cli/cli.py [hunk 12]
```diff
@@ -46,66 +46,52 @@
...
+
+    config_str = render.format_object_as_yaml(searcher_config)
+    output.append(config_str)
+    headers = ["Runs", "Training Time"]
+    run_summaries = []
+    for run_summary in resp.summary.runs:
+        num_runs = run_summary.count
+        run_unit = run_summary.unit
+        if run_unit.maxLength:
+            summary = "maximum length of training code"
```
Comment (rb-determined-ai) [inline harness/determined/cli/cli.py:71]: I feel like this phrasing is weird.  Do you like "Training time: to completion" better than "Training time: maximum length of training code"?

I think it's the "training **TIME** = maximum **LENGTH**" that feels like somebody is answering a different question than what they were asked, or something.

File: harness/tests/cli/util.py [hunk 13]
```diff
@@ -151,3 +154,18 @@ def expect_get_info(
...
+    matches expected output.
+    """
+    with contextlib.redirect_stdout(io.StringIO()) as f:
+        cli.main(args=args)
+    actual = f.getvalue()
+    exp_lines = expected.splitlines(keepends=True)
+    act_lines = actual.splitlines(keepends=True)
+    diff_lines = difflib.ndiff(act_lines, exp_lines)
+    diff = "".join(diff_lines)
+    assert expected == actual, f"CLI output for {args} does not match expected output:\n {diff}"
```
Comment (rb-determined-ai) [inline harness/tests/cli/util.py:171]: You dropped the indicators from my suggestions to help parse what was actual and what was received.  I had:

```python
raise ValueError(f"cli output for {args} output actual (-) != expected (+):\n{diff}")
```

Because that tells you if you see a `-` line, that means it was in actual but not expected, and if you see a `+` line it was in expected but not actual.

File: master/internal/restore.go [hunk 14]
```diff
@@ -134,79 +133,68 @@ func (m *Master) restoreExperiment(expModel *model.Experiment) error {
...
 }
 
-// restoreTrial takes the a searcher.Create and attempts to restore the trial that would be
-// associated with it. On failure, the trial is just reset to the start and errors are logged.
-func (e *internalExperiment) restoreTrial(
-	ckpt *model.Checkpoint, searcher experiment.TrialSearcherState,
+func (e *internalExperiment) restoreRun(
+	ckpt *model.Checkpoint, searcher experiment.RunSearcherState,
 ) {
-	l := e.syslog.WithField("request-id", searcher.Create.RequestID)
```
Comment (rb-determined-ai) [inline master/internal/restore.go:142]: why not include run-id as a field in your log here?

File: master/internal/restore.go [hunk 15]
```diff
@@ -134,79 +133,68 @@ func (m *Master) restoreExperiment(expModel *model.Experiment) error {
...
+			syslog.WithError(err).Error("failed to retrieve previous run, restoring new run")
+		} else {
+			syslog = syslog.WithField("run-id", trial.ID)
 		}
 	}
 
-	taskID := trialTaskID(e.ID, searcher.Create.RequestID)
-	if !terminal && trialID != nil {
-		trialTaskIDs, err := db.TrialTaskIDsByTrialID(context.TODO(), *trialID)
+	taskID := model.TaskID(fmt.Sprintf("%d.%s", e.ID, model.NewTaskID()))
```
Comment (rb-determined-ai) [inline master/internal/restore.go:159]: So we get a new _task_ for a restored trial, not just a new allocation of the same task?

File: master/internal/restore.go [hunk 16]
```diff
@@ -444,3 +445,180 @@ func shimExperimentSnapshotV4(snapshot []byte) ([]byte, error) {
 
 	return json.Marshal(experimentSnapshotV4)
 }
+
+// shimExperimentSnapshotV5 shims a v4 snapshot to a v5 snapshot. From v4 to v5:
```
Comment (rb-determined-ai) [inline master/internal/restore.go:449]: copy paste errors in this line

File: master/internal/restore.go [hunk 17]
```diff
@@ -265,7 +254,7 @@ func shimSnapshot(shims map[int]snapshotShimFunc, snapshot []byte, version int)
 		if !ok {
 			return nil, fmt.Errorf("missing shim from %d to %d", version, experimentSnapshotVersion)
```
Comment (rb-determined-ai) [inline master/internal/restore.go:255]: nit, this should be `"from %d to %d", version, version+1`

File: master/internal/restore.go [hunk 18]
```diff
@@ -246,16 +234,17 @@ var experimentSnapshotShims = map[int]snapshotShimFunc{
...
 }
 
 // shimExperimentSnapshot shims an experiment snapshot to the version required by the master,
 // returning an error in the event the shim fails or the snapshot version is greater
 // than the current version (which could happen in a downgrade).
-func shimExperimentSnapshot(snapshot []byte, version int) ([]byte, error) {
-	return shimSnapshot(experimentSnapshotShims, snapshot, version)
+func shimExperimentSnapshot(experimentID int, snapshot []byte, version int) ([]byte, error) {
+	return shimSnapshot(experimentID, experimentSnapshotShims, snapshot, version)
 }
```
Comment (rb-determined-ai) [inline master/internal/restore.go:245]: question: why the extra helper function?  Why not just `shimSnapshot` directly, and reference `experimentShapshotShims` directly?

Kinda seems like there were dreams of having a generic shimming tool, but that seems kinda pointless, and this is just extra layers of indirection.

File: master/internal/restore.go [hunk 19]
```diff
@@ -344,11 +333,23 @@ func shimExperimentSnapshotV1(snapshot []byte) ([]byte, error) {
 
 // Version 2 => 3 shims
 
+// All the operation types that support serialization.
```
Comment (rb-determined-ai) [inline master/internal/restore.go:336]: Should this comment be something like:

```go
// Old serialization values which are no longer a part of the searcher package.
```
?

File: master/pkg/model/experiment.go [hunk 20]
```diff
@@ -568,7 +567,7 @@ func NewTrial(
 		warmStartCheckpointID = &warmStartCheckpoint.ID
 	}
 	return &Trial{
-		RequestID:             &requestID,
+		RequestID:             nil,
```
Comment (rb-determined-ai) [inline master/pkg/model/experiment.go:570]: can you delete this field yet?  Or do you need to leave it in the DB for restorability reasons, but for new trials it is always nil?

File: master/pkg/schemas/expconf/searcher_config.go [hunk 21]
```diff
@@ -247,9 +218,47 @@ type CustomConfigV0 struct {
...
+					"future release.")
+		}
+		if s.RawAdaptiveASHAConfig.RawStopOnce != nil {
+			log.Warn("the `stop_once` field of the searcher config has been deprecated and will be removed in " +
+				"a future release.")
+		}
+		if s.RawAdaptiveASHAConfig.RawMaxTime == nil || s.RawAdaptiveASHAConfig.RawTimeMetric == nil {
+			return errors.New("the `adaptive_asha` searcher requires `max_time` and `time_metric` to be set")
+		}
+	case s.RawAsyncHalvingConfig != nil:
```
Comment (rb-determined-ai) [inline master/pkg/schemas/expconf/searcher_config.go:238]: doesn't AsyncHalving require the same max_time and time_metric fields?

File: master/internal/restore.go [hunk 22]
```diff
@@ -134,79 +133,68 @@ func (m *Master) restoreExperiment(expModel *model.Experiment) error {
...
-		terminal = true
-	default:
-		trialID = &trial.ID
-		l = l.WithField("trial-id", trial.ID)
-		if model.TerminalStates[trial.State] {
-			l.Debugf("trial was in terminal state in restore: %s", trial.State)
-			terminal = true
+	var err error
+
+	if searcher.RunID != nil {
```
Comment (rb-determined-ai) [inline master/internal/restore.go:146]: Something doesn't smell right here.

- Why is the runid ever nil on the snapshot?
- Assuming the previous answer is "b/c ordering is hard", then why do we save said snapshot to the db?
- Assuming there is a good answer to that, then why do we create a new trial as part of restoring?  If there was a good reason to have a nil RunID before the shutdown, then why doesn't that reason still apply during restore?

**Review by rb-determined-ai — COMMENTED** [review]
oh, I have a question that has no home.

Was there a plan to have kill/cancel/cancel-with-timeout behavior for the searcher as part of this work?

File: master/internal/restore.go (hunk 22 above)
Comment (azhou-determined) [reply master/internal/restore.go:146]: yeah, this was confusing to me as well. i was trying my best to refactor existing code while keeping the same behavior. AFAICT, the old code _was_ creating a new trial if the snapshot's trial wasn't created. 

in previous `restore.go`, we try to get trial by request ID. if it doesn't exist, we create a new trial. in `experiment.go`, searcher wants to create trial -> we set the trial state -> create trial -> snapshot. but i agree, i can't imagine why you'd want to recreate the trial if something fails before snapshot. 

maybe this behavior is just a leftover artifact of some refactored behavior, like if searcher create -> snapshot used to be a more async process. happy to get rid of it, but maybe @stoksc can confirm here?

File: master/internal/restore.go (hunk 14 above)
Comment (azhou-determined) [reply master/internal/restore.go:142]: because it could be nil at this point, a few lines down there's a nil check and if it exists, it is set on the syslog. 

question of whether it could be nil at this point is in the other comment

File: master/pkg/model/experiment.go (hunk 20 above)
Comment (azhou-determined) [reply master/pkg/model/experiment.go:570]: yeah, restore logic queries the old trial's request id.

**Comment by azhou-determined** [comment]
> oh, I have a question that has no home.
> 
> Was there a plan to have kill/cancel/cancel-with-timeout behavior for the searcher as part of this work?

this was done separately. `preemption_timeout` is a configuration option in the expconf, which applies to terminating all tasks, including when searcher early-stops runs.

File: master/internal/restore.go [hunk 23]
```diff
@@ -134,79 +133,68 @@ func (m *Master) restoreExperiment(expModel *model.Experiment) error {
...
+				e.trialExited(*searcher.TrialID, nil)
+			}
+			return
+		}
+		trialTaskIDs, err := db.TrialTaskIDsByTrialID(context.TODO(), trial.ID)
 		switch {
 		case err != nil:
-			l.WithError(err).Error("failed to retrieve trial's tasks, stopping restore")
-			terminal = true
+			syslog.WithError(err).Error("failed to retrieve tasks for run")
```
Comment (stoksc) [inline master/internal/restore.go:174]: we removed the `terminal` bit. i think we need to add it back and decide to close the searcher in some of these cases where we don't have enough information to continue.

File: master/internal/trial.go [hunk 24]
```diff
@@ -230,13 +230,22 @@ func (t *trial) PatchSearcherState(req experiment.TrialSearcherState) error {
...
+				InformationalReason: "searcher decided to early stop trial",
+			},
+		)
+	case t.searcher.Closed:
 		return t.patchState(model.StateWithReason{
 			State:               model.StoppingCompletedState,
 			InformationalReason: "hp search is finished",
 		})
+	default:
+		if !t.searcher.Stopped && !t.searcher.Closed {
```
Comment (stoksc) [inline master/internal/trial.go:246]: isn't this a tautology?

File: master/pkg/searcher/actions.go [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	"fmt"
+
+	"github.com/determined-ai/determined/master/pkg/nprand"
+)
+
+// Action is an action that a searcher would like to perform.
+type Action interface {
+	String() string
```
Comment (stoksc) [inline master/pkg/searcher/actions.go:11]: if you want a disciminated union, you can use the same trick i did with `sproto.ResourceEvent`

File: master/pkg/searcher/asha_stopping.go [hunk 26]
```diff
@@ -112,111 +132,113 @@ func (s *asyncHalvingStoppingSearch) initialOperations(ctx context) ([]Operation
...
 
-	allTrials := len(s.TrialRungs) - s.InvalidTrials
-	if !addedTrainWorkload && allTrials < s.MaxTrials() {
-		create := NewCreate(
-			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
-		s.TrialRungs[create.RequestID] = 0
-		ops = append(ops, create)
-		ops = append(ops, NewValidateAfter(create.RequestID, s.Rungs[0].UnitsNeeded))
-	}
+		// Top 1/divisor trials should continue, trials - 1/divisor trials should be stopped.
```
Comment (stoksc) [inline master/pkg/searcher/asha_stopping.go:223]: comment nit, 1/divisor is a percent of trials so `trials - 1/divisor` doesn't really make sense.

File: master/pkg/searcher/asha_stopping.go [hunk 27]
```diff
@@ -112,111 +132,113 @@ func (s *asyncHalvingStoppingSearch) initialOperations(ctx context) ([]Operation
...
-			trialMetric{
-				RequestID: requestID,
-				Metric:    model.ExtendedFloat64(metric),
-			},
-		)
+// stopRun handles early-stopping and record-keeping logic for a validation metric reported to the
+// searcher.
+// If the metric qualifies the run for a rung but is not in the top 1/divisor trials for that rung,
+// stopRun will return a single `searcher.Stop` action. Otherwise, no actions will be returned.
+func (s *asyncHalvingStoppingSearch) stopRun(
```
Comment (stoksc) [inline master/pkg/searcher/asha_stopping.go:198]: i would call this doEarlyStopping or something. stopRun is misleading.

File: master/pkg/searcher/asha_stopping.go [hunk 28]
```diff
@@ -112,111 +132,113 @@ func (s *asyncHalvingStoppingSearch) initialOperations(ctx context) ([]Operation
...
-	if !addedTrainWorkload && allTrials < s.MaxTrials() {
-		create := NewCreate(
-			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
-		s.TrialRungs[create.RequestID] = 0
-		ops = append(ops, create)
-		ops = append(ops, NewValidateAfter(create.RequestID, s.Rungs[0].UnitsNeeded))
-	}
+		// Top 1/divisor trials should continue, trials - 1/divisor trials should be stopped.
+		// If trials < divisor, continue only if this is the best performing run so far.
+		numContinue := mathx.Max(int(float64(len(rung.Metrics))/s.Divisor()), 1)
```
Comment (stoksc) [inline master/pkg/searcher/asha_stopping.go:225]: why is the original algorithm was
```suggestion
		numContinue := mathx.Max(int(float64(len(rung.Metrics) + 1)/s.Divisor()), 1)
```
and yours drops the + 1? i guess the max protects us from an aggressive divisor not promoting anything to the last rung. surely liam would know? some weird floating point protection, haha?

Comment (azhou-determined) [reply master/pkg/searcher/asha_stopping.go:225]: i dropped it simply because before, this was calculated before the new metric was added, hence the +1 to the metric count.

Comment (stoksc) [reply master/pkg/searcher/asha_stopping.go:225]: that may be all it is haha. i just was reading the new code as careful as i could for any weird differences.

File: master/internal/api_trials.go [hunk 29]
```diff
@@ -1499,20 +1431,19 @@ func (a *apiServer) ReportTrialProgress(
...
+	if err := e.RunReportProgress(req.TrialId, msg); err != nil {
 		return nil, err
 	}
 	return &apiv1.ReportTrialProgressResponse{}, nil
 }
 
 func (a *apiServer) ReportTrialMetrics(
 	ctx context.Context, req *apiv1.ReportTrialMetricsRequest,
-) (*apiv1.ReportTrialMetricsResponse, error) {
+) (resp *apiv1.ReportTrialMetricsResponse, err error) {
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply master/internal/api_trials.go]: nah, changed

File: master/pkg/searcher/actions.go (hunk 25 above)
Comment (stoksc) [reply master/pkg/searcher/actions.go]: yeah, no typing.Unions here

File: master/internal/experiment.go [hunk 30]
```diff
@@ -321,8 +321,8 @@ func (e *internalExperiment) TrialReportProgress(trialID int32, msg experiment.T
 
 	progress := float64(msg.Progress)
 	e.searcher.SetTrialProgress(trialID, progress)
-
-	if err := e.db.SaveExperimentProgress(e.ID, &progress); err != nil {
+	experimentProgress := e.searcher.Progress()
```
Comment (stoksc) [inline master/internal/experiment.go:324]: oof good catch

File: master/pkg/searcher/actions.go [hunk 31]
```diff
@@ -0,0 +1,73 @@
...
+	"fmt"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+
+	"github.com/determined-ai/determined/master/pkg/nprand"
+)
+
+// Action is an action that a searcher would like to perform.
+type Action interface {
+	SearcherAction()
```
Comment (stoksc) [inline master/pkg/searcher/actions.go:13]: i think this can be unexported
```suggestion
	searcherAction()
```

File: master/internal/experiment.go [hunk 32]
```diff
@@ -699,73 +621,57 @@ func (e *internalExperiment) processOperations(
...
+				e.trialExited(action.RequestID, ptrs.Ptr(model.Errored))
 				continue
 			}
 
 			t, err := newTrial(
-				e.logCtx, trialTaskID(e.ID, op.RequestID), e.JobID, e.StartTime, e.ID, e.State,
-				state, e.rm, e.db, config, checkpoint, clonedSpec, e.generatedKeys, false,
-				nil, continueFromTrialID, e.TrialClosed,
+				e.logCtx, trialTaskID(e.ID, action.RequestID), e.JobID, e.StartTime, e.ID, e.State,
+				state, e.rm, e.db, config, e.warmStartCheckpoint, clonedSpec, e.generatedKeys, false,
```
Comment (stoksc) [inline master/internal/experiment.go:652]: i don't quite follow why we removed `checkpointForCreate` and the work it did. i tried to hold opinions until i read `trial.go` but i dont see any logic moved there.

does python code do it now or something?

File: master/internal/api_trials.go [hunk 33]
```diff
@@ -1528,6 +1454,22 @@ func (a *apiServer) ReportTrialMetrics(
...
 		return nil, err
 	}
+	if metricGroup == model.ValidationMetricGroup {
+		// Notify searcher of validation metrics.
+		eID, rID, err := a.m.db.TrialExperimentAndRequestID(int(req.Metrics.TrialId))
+		if err != nil {
+			return nil, errors.Errorf("Failed to get experiment ID from trial ID (%d)", req.Metrics.TrialId)
+		}
+		e, ok := experiment.ExperimentRegistry.Load(eID)
+		if ok {
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply master/internal/api_trials.go:1464]: nah, Load will be !ok if key doesn't exist

[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/api_trials.go:1464]: sorry mt; i wondered if it _should_ explode.

File: master/internal/db/postgres_trial.go [hunk 34]
```diff
@@ -276,17 +276,28 @@ func TrialTaskIDsByTrialID(ctx context.Context, trialID int) ([]*model.RunTaskID
...
+// TrialIDsByExperimentIDAndRequestIDs looks up trial IDs by experiment ID and request IDs, returning an error if
+// any fail. This is only used to shim legacy experiment snapshots.
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply master/internal/db/postgres_trial.go]: whole method is stale, removed

File: master/internal/experiment.go (hunk 32 above)
Comment (azhou-determined) [reply master/internal/experiment.go:652]: it wasn't being used AFAICT, pretty sure we needed it at some point to support PBT and custom searcher, but both of those are dead now.

File: master/pkg/searcher/simulate.go [hunk 35]
```diff
@@ -1,222 +1,121 @@
...
-			pending[op.RequestID] = []Operation{op}
-		case Requested:
-			pending[op.GetRequestID()] = append(pending[op.GetRequestID()], op)
-		case Shutdown:
-			return true, nil
-		default:
-			return false, errors.Errorf("unexpected operation: %T", operation)
-		}
+// Simulate generates the intended training plan for the searcher.
+func Simulate(conf expconf.SearcherConfig, hparams expconf.Hyperparameters) (SearchSummary, error) {
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply master/pkg/searcher/simulate.go:65]: well, ASHA is no longer controlled by the searcher sending `ValidateAfter` workloads, so there was no easy way to implement this same behavior. like, it's up to users to validate at specified intervals and the searcher will try and roughly stop on rung boundaries. 

i guess this simulate is less rigorous because the implementation is less rigorous.

File: master/internal/trial.go [hunk 36]
```diff
@@ -230,15 +230,21 @@ func (t *trial) PatchSearcherState(req experiment.TrialSearcherState) error {
 
 	t.searcher = req
 	switch {
-	case !t.searcher.Complete:
-		return t.maybeAllocateTask()
-	case t.searcher.Complete && t.searcher.Closed:
+	case t.searcher.Stopped:
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply master/internal/trial.go:233]: yeah i agree it's a bit confusing. right now, stopped = searcher decided to early stop trial, closed = trial errored or was canceled. what about stopped / exited?

**Review by stoksc — APPROVED** [review]
lgtm pending that last name change for closed/stopped

**Review by stoksc — APPROVED** [review]
